File: ruby/4b.rb
#!/usr/bin/ruby

rooms = File.open('4.in').readlines.map do |x|

  # Extract the room
  r = x.rpartition('-').first

  # Group by characters (frequency count), sort by freqency and alphabet
  c = r.gsub(/-/, '').chars.group_by(&:chr).sort_by{ |x,y| [0 - y.length, x] }

  # Return an array containing computed checksum, sector and checksum
  [ c.map(&:first)[0..4].join, r, x.scan(/\d+/).first.to_i, x.scan(/\w+/).last]

end.select{ |x| x.first == x.last }

rooms.each do |x|

  # Convert dashes to spaces and rotate characters caesar style
  plaintext = x[1].chars.map{ |y| y == '-' ? ' ' : ((y.ord - 97 + x[2]).modulo(26) + 97).chr }.join

  # Look for the north pole storage room
  next unless plaintext.include?('north')

  # Terminate when complete
  puts x[2]
  exit

end

File: python/4b.py
#!/usr/bin/python

import collections

def main():
    acc = []
    for line in open('4.in'):
        # Extract the room and extra data
        room, extra = line.rsplit('-', 1)
        # Count the occurrences of letters ignoring hyphens
        counts = collections.Counter(room.replace('-', ''))
        # Sort based on highest requency then lexically
        ordered = sorted(counts.items(), key=lambda x: (-x[1], x[0]))
        # The checksum is the first 5 characters from our ordered list
        computed = ''.join(x for x, y in ordered[:5])
        # Extract the sector and checksum
        sector, checksum = extra[:-2].split('[')
        # Ignore faulty checksums
        if checksum != computed:
            continue
        # Accumulate room and sector
        acc.append((room, int(sector)))
    for room, sector in acc:
        # Decipher the room, replacing hyphens with spaces
        def caesar(char, num):
            offset = ord('a')
            return chr((ord(char) - offset + num) % 26 + offset)
        plain = ''.join(caesar(x, sector) for x in room.replace('-', ' '))
        # Is this the north pole room?
        if 'north' in plain:
            print sector
            break

if __name__ == '__main__':
    main()

# vi: ts=4 et:
